vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -832,16 +832,12 @@ func resourceVSphereVirtualMachineCreate(d *schema.ResourceData, meta interface{
 				}
 
 				if v, ok := disk["size"].(int); ok && v != 0 {
-					if v, ok := disk["template"].(string); ok && v != "" {
-						return fmt.Errorf("Cannot specify size of a template")
-					}
 
 					if v, ok := disk["name"].(string); ok && v != "" {
 						newDisk.name = v
-					} else {
+					} else if v, ok := disk["template"].(string); !ok && v != "" {
 						return fmt.Errorf("[ERROR] Disk name must be provided when creating a new disk")
```

Comment:
Is this error message still correct?

Reply:
Prior to change, we check to make sure `name` for the disk is provided and fail if not. With the change, we check for the `name` but only fail where disk is not created from a template. Prior to the change every condition where "disk/size" exist required "disk/name" to be present. Now "disk/size" can be used without "disk/name" if disk is a template.

I can change the error string to something like `Disk name must be provided when creating a new disk that is not a template`.

Reply:
Hey @alekssaul, I'm kind of wondering if maybe @grubernaut was asking if maybe the error _condition_ was not correct here.

It's checking if the the value of a `map[string]interface{}` is a valid string, I'm assuming, but the conditional fires if:

* The key is not present (`!ok`)
* **and if only also if** `v` is not empty (`&& v != ""`)

Since the zero value of a map[string]interface is `nil`, the value of `v` when `!ok` is true will always be `nil`.

To have the condition match the message, I think you would want:

```
if v, ok := disk["template"].(string); !ok || v == ""
```

This condition would be true if the key was not present OR if `v` was an empty string.